src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php

```diff
@@ -0,0 +1,199 @@
+    {
+        return $this->client->getMulti($ids);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function doHave($id)
+    {
+        return $this->client->get($id) !== false
+            || $this->client->getResultCode() === \Memcached::RES_SUCCESS;
+    }
+    /**
+     * {@inheritdoc}
+     */
+    protected function doDelete(array $ids)
+    {
+        $toDelete = count($ids);
+        foreach ((array) $this->client->deleteMulti($ids) as $result) {
+            if (true === $result || \Memcached::RES_NOTFOUND === $result) {
+                --$toDelete;
+            }
+        }
+
+        return 0 === $toDelete;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function doClear($namespace)
+    {
+        if (!isset($namespace[0]) || false === $ids = $this->getIdsByPrefix($namespace)) {
+            return $this->client->flush();
+        }
+
+        $isCleared = true;
+        do {
+            $isCleared = $this->doDelete($ids) && $isCleared;
+        } while ($ids = $this->getIdsByPrefix($namespace));
+
+        return $isCleared;
+    }
+
+    private function getIdsByPrefix($namespace)
+    {
+        if (false === $ids = $this->client->getAllKeys()) {
+            return false;
+        }
+
+        return array_filter((array) $ids, function ($id) use ($namespace) {
+            return 0 === strpos($id, $namespace);
+        });
+    }
+
+    private function setOption($opt, $val)
+    {
+        $toRestore = error_reporting(~E_ALL);
+        $isSuccess = $this->client->setOption($this->resolveOption($opt), $this->resolveOption($val));
+        error_reporting($toRestore);
+
+        return $isSuccess;
+    }
+
+    private function resolveOption($val)
+    {
+        return defined($constant = '\Memcached::'.strtoupper($val)) ? constant($constant) : $val;
+    }
+
+    private function addServer($dsn)
+    {
+        list($host, $port, $weight) = $this->parseServerDSN($dsn);
+
+        return $this->isServerInClientPool($host, $port)
+            || ($this->client->addServer($host, $port, $weight)
+                && $this->client->getResultCode() === \Memcached::RES_SUCCESS);
+    }
+
+    private function parseServerDSN($dsn)
+    {
+        if (false === ($srv = parse_url($dsn)) || $srv['scheme'] !== 'memcached' || count($srv) > 4) {
+            throw new InvalidArgumentException(sprintf('Invalid Memcached DSN: %s (expected "memcached://[<string>[:<int>]][?weight=<int>]")', $dsn));
+        }
+
+        array_shift($srv);
+        $srv += self::$defaultClientServer;
+
+        if (isset($srv['query']) && 1 === preg_match('{weight=([^&]{1,})}', $srv['query'], $weight)) {
+            unset($srv['query']);
+            $srv['weight'] = (int) $weight[1];
+        }
+
+        return array_values($srv);
+    }
+
+    protected function isServerInClientPool($host, $port)
+    {
+        return (bool) array_filter($this->client->getServerList(), function ($srv) use ($host, $port) {
```

Comment:
using array_filter is a bad idea: it forces to compare all servers even if the first one is a match, meaning that you force to always reach the worse case (when there is no match, we are indeed forced to compare them all)